### PR TITLE
Add tuning support for FPU reduction

### PR DESCRIPTION
**Pull request by jkiliani**
Change match default setting from --noise to --tempdecay 10

### CODE REVIEW

**Comment by jkiliani** [comment]
OK this went a bit wrong... seems the change to the match server was done without actually merging this PR? Now I added the addition of FPU reduction to the tuner accidentally...

**Comment by killerducky** [comment]
The tempdecay change was made on master, but not on next. I have merged master back into next:
https://github.com/glinscott/leela-chess/commit/dfb6342ae5cc3167ff085af24459f8e8fc6dbd0e

I'm starting to think having separate master/next branches might be causing more problems than it's worth. Maybe we should switch back to only a master branch? If we ever get into a situation where we have to go back to e.g. released v0.5 and hotfix it we can create a special hotfix branch for that. Any complicated changes can be worked in a feature branch, e.g. what we did with the hist_fix branch.

@glinscott @Error323 what do you think?

**Comment by mooskagh** [comment]
Usually master is the branch for development. When new version is about to
release, branch or tag is created for that release, and if some hotfixes
are needed later, they are cherrypicked.

Actually it's almost like it is now, just next has to be renamed into
master and master into release (or per-version branch).

пт, 13 кра 2018, 17:35 карыстальнік Andy Olsen <notifications@github.com>
напісаў:

File: src/main.cpp [hunk 1]
```diff
@@ -90,6 +90,7 @@ static std::string parse_commandline(int argc, char *argv[]) {
 #endif
 #ifdef USE_TUNER
         ("puct", po::value<float>())
+        ("fpu_reduction", po::value<float>())
```
Comment (Tilps) [inline src/main.cpp]: Consider making dynamic eval enable/disable at the same time?

Comment (jkiliani) [reply src/main.cpp]: Good idea, I'll put it in tomorrow since I can't code at the moment.

**Comment by jkiliani** [comment]
It just occurred to me, is there a good reason to keep USE_TUNER deactivated by default? Client.exe passes on the correct set of parameters for self-play and matches anyway, so why not just let people on precompiled binaries tune by default, if they run Leela against other engines or humans? Does anyone see something problematic with that?

**Comment by killerducky** [comment]
@jkiliani can you please add a fix for #317 to this?

**Comment by jkiliani** [comment]
@killerducky OK I think I made a version of get_eval that only deactivated virtual losses. Some sample control output for the first move looks like this:
```
Dynamic node eval with VL:  0.52418 without VL:  0.52418
Dynamic node eval with VL:  0.41716 without VL:  0.51342
Dynamic node eval with VL:  0.33822 without VL:  0.50734
Dynamic node eval with VL:  0.50370 without VL:  0.50370
Dynamic node eval with VL:  0.32326 without VL:  0.51722
Dynamic node eval with VL:  0.49538 without VL:  0.49538
Dynamic node eval with VL:  0.50178 without VL:  0.50178
```
I scrolled though a number of node evals, and it seems the fix always returns stable evals around 0.5, while the previous dynamic eval is identical in most cases and significantly reduced in some, but never higher. This looks to me as if it works, but I can't guarantee it, feel free to review and test yourself.

**Comment by jkiliani** [comment]
I added search depth tracking to this branch, and also reverted the use of get_raw_eval for now, since it looks unlikely we could easily find another FPU reduction method as strong as that provided by virtual losses.

**Comment by jkiliani** [comment]
Rebased, closing this for #364 .

**Pull request closed without merging**